iron-form.html

```diff
@@ -0,0 +1,146 @@
+-->
+
+<script>
+
+  Polymer({
+
+    is: 'iron-form',
+
+    extends: 'form',
+
+    /**
+     * Fired after the form is submitted.
+     *
+     * @event iron-form-submit
+     */
+
+    /**
+    * Fired after the form is submitted and a response is received.
+    *
+    * @event iron-form-response
+    */
+
+    /**
+     * Fired after the form is submitted and an error is received.
+     *
+     * @event iron-form-error
+     */
+    properties: {
+
+      /**
+       * Contains all the registered form elements and their values that
+       * will be used to submit the form.
+       */
+      jsonData: {
+        type: Object,
+        notify: true
+      }
+
+    },
+
+    listeners: {
+      'iron-form-element-register': '_registerElement',
+      'submit': 'submit'
+    },
+
+    ready: function() {
+      // Object that handles the ajax form submission request.
+      this._requestBot = document.createElement('iron-ajax');
+      this._requestBot.addEventListener('response', this._handleFormResponse.bind(this));
+      this._requestBot.addEventListener('error', this._handleFormError.bind(this));
+
+      // Holds all the custom elements registered with this form.
+      this._customElements = [];
+    },
+
+    /**
+     * Called to submit the form.
+     */
+    submit: function(event) {
+      var valid = true;
+
+      for (var el, i = 0; el = this._customElements[i], i < this._customElements.length; i++) {
+        valid = el.validate() && valid;
+      }
+
+      if (!valid) {
+        return false;
+      }
+
+      this.serialize();
+
+      this._requestBot.url = this.action;
+      this._requestBot.method = this.method;
+      this._requestBot.params = this.jsonData;
+
+      if (this.method == 'POST') {
+        this._requestBot.body = JSON.stringify(this.jsonData);
+      }
+
+      this._requestBot.generateRequest();
+      this.fire('iron-form-submit');
+
+      // Don't perform a page refresh.
+      if (event)
+        event.preventDefault();
+
+      return false;
+    },
+
+    /**
+     * Serializes all the registered custom components and native elements of
+     * the form. The aggregated json object is stored in `jsonData`.
+     */
+    serialize: function() {
+      var json = {};
+
+      // Go through all of the registered custom components.
+      for (var el, i = 0; el = this._customElements[i], i < this._customElements.length; i++) {
+        if (el.name) {
+          json[el.name] = el.value;
+        }
+      }
+
+      // Also go through the form's native elements.
+      for (var el, i = 0; el = this.elements[i], i < this.elements.length; i++) {
+        // Don't re-add custom elements that extend native elements (like an
+        // `<input is="fancy-input">`), since they will also show up in this list.
+        if (el.name && !json[el.name]) {
+          json[el.name] = el.value;
+        }
+      }
+
+      this.jsonData = json;
```

Comment:
It seems potentially weird that `jsonData` is a persistent property of the form, but its values are likely to fall out-of-sync with the values of the actual form inputs. Is it useful or necessary for `jsonData` to be a property?